topic-operator/src/main/java/io/strimzi/operator/topic/TopicSerialization.java

```diff
@@ -45,52 +50,82 @@
-            for (Map.Entry<String, Object> entry : kafkaTopic.getSpec().getConfig().entrySet()) {
-                String key = entry.getKey();
-                Object v = entry.getValue();
-                boolean isNumberType = v instanceof Long
-                        || v instanceof Integer
-                        || v instanceof Short
-                        || v instanceof Double
-                        || v instanceof Float;
-                if (v instanceof String
-                        || isNumberType
-                        || v instanceof Boolean) {
-                    result.put(key, v.toString());
-                } else {
-                    String msg = "The value corresponding to the key must have a string, number or boolean value";
-                    if (v == null) {
-                        msg += " but the value was null";
-                    } else {
-                        msg += " but was of type " + v.getClass().getName();
-                    }
-                    throw new InvalidTopicException(kafkaTopic, "KafkaTopic's spec.config has invalid entry: " +
-                            "The key '" + key + "' of the topic config is invalid: " + msg);
+    /**
+     * Gets the config model for the given version of the Kafka broker.
+     * @param kafkaVersion The broker version.
+     * @return The topic config model for that broker version.
+     */
+    private static Map<String, ConfigModel> readConfigModel(String kafkaVersion) {
+        String name = "/kafka-" + kafkaVersion + "-topic-config-model.json";
+        try {
+            try (InputStream in = TopicSerialization.class.getResourceAsStream(name)) {
+                ConfigModels configModels = new ObjectMapper().readValue(in, ConfigModels.class);
+                if (!kafkaVersion.equals(configModels.getVersion())) {
+                    throw new RuntimeException("Incorrect version");
                 }
+                return configModels.getConfigs();
+            }
+        } catch (IOException e) {
+            throw new RuntimeException("Error reading from classpath resource " + name, e);
+        }
+    }
+
+    /**
+     * Validate the configs in this KafkaTopicConfiguration returning a list of errors.
+     * @param topicConfiguration The topic configuration.
+     * @param kafkaVersion The broker version.
+     * @return A list of error messages.
+     */
+    private static List<String> validate(Map<String, String> topicConfiguration, String kafkaVersion) {
+        List<String> errors = new ArrayList<>();
+        Map<String, ConfigModel> models = readConfigModel(kafkaVersion);
+        for (Map.Entry<String, String> entry: topicConfiguration.entrySet()) {
+            String key = entry.getKey();
+            String value = entry.getValue();
+            ConfigModel config = models.get(key);
+            if (config != null) {
+                errors.addAll(config.validate(key, value));
```

Comment:
Do we need to keep the type coercion code, in order to avoid regression? Or are we expecting the bidirectionality means that previously even if the user's initial CR used a type that got coerced, it would have been converted to some canonical type following reconciliation?

Reply:
The coercion of type is preserved.